### ISSUE TITLE
`func-name-matching`: "never"

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#7391]
#Change request: `func-name-matching`

This rule currently has no options, and requires that function names always match the variable or property to which they are assigned.

I'd like to request the schema be expanded to accept a string: "always" (the default) or "never" (the new one).

"never" would require that the function name *differ* from the variable or property to which it is assigned.

Per https://github.com/eslint/eslint/pull/7063#issuecomment-247104931 / https://github.com/eslint/eslint/pull/7063#issuecomment-247106545

I'll be happy to make a PR for this.
-----
*Updated*: answers from http://eslint.org/docs/developer-guide/contributing/changes

> The version of ESLint you are using.

v3.8.0

> The problem you want to solve.

I want to prevent named functions from being stored in variables or properties that match their name - in other words, to force the function name and the container name to be different.

> Your take on the correct solution to problem.

An optional string (before the object) that defaults to "always" when omitted, and can be set to "never".

> If you’re requesting a change to a rule, it’s helpful to include this information as well:
> What you did.

```js
const foo = function foo() {};
const bar = function baz() {};
```

> What you would like to happen.

A warning on the first line only.

> What actually happened.

A warning on the second line only (with `func-name-matching` enabled).

**Comment by ljharb** [comment]
Actually it seems that the rule accepts a "includeCommonJSModuleExports" option - so instead of accepting a string, I'd propose another object property option, "nameMatches" - which defaults to "always", with the new option being "never".

**Comment by platinumazure** [comment]
Two questions:

1. If you use the "never" option but then fail to name the function expression at all, should that still warn? (That is, is the name required even with the new option?)
1. What about arrow functions, should they be reported at all here?

And if you wouldn't mind filling out the Rule Change Proposal form (link forthcoming), that would be awesome. Thanks!

**Comment by eslintbot** [comment]
Hi @ljharb, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ljharb** [comment]
@platinumazure
1. no, the `func-names` rule covers that case.
2. arrow functions don't have names, so it wouldn't make sense for a rule about function names to care about them.

Thanks! I couldn't find the info last night :-) I'll update the OP with that info.

**Comment by ljharb** [comment]
@platinumazure updated.

**Comment by platinumazure** [comment]
Thanks @ljharb! I will champion this change.

**Comment by ljharb** [comment]
@platinumazure Ready in #7428

**Comment by platinumazure** [comment]
@eslint/eslint-team Still need one more :+1: for this change!

**Comment by mikesherov** [comment]
:+1:
